### PR TITLE
Change platform API to include `init` and `respond`

### PR DESCRIPTION
**Pull request by lukewilliamboswell**

### CODE REVIEW

File: crates/roc_host/src/roc.rs [hunk 1]
```diff
@@ -945,3 +886,89 @@ pub extern "C" fn roc_fx_tempDir() -> RocList<u8> {
...
+            captures.as_mut_ptr(),
+            model.as_mut_ptr(),
+        );
+
+        model.set_readonly();
+        captures.set_readonly();
+
+        RocServer {
+            model: model.into(),
+            captures: captures.into(),
```
Comment (bhansconnect) [inline crates/roc_host/src/roc.rs:930]: You can't save captures and keep reusing it like this. It isn't valid. This only works for `model` cause `model` is a `Box something`.

You need to do the same thing that we do in wasm4: https://github.com/lukewilliamboswell/roc-wasm4/blob/a216eb96e5d05dc2258cbf3da74facf787e93bcc/platform/host.zig#L133-L137

File: crates/roc_host/src/roc.rs [hunk 2]
```diff
@@ -945,3 +886,89 @@ pub extern "C" fn roc_fx_tempDir() -> RocList<u8> {
...
+        fn roc__forHost_1_exposed_generic(_: *mut u8);
+        fn roc__forHost_1_exposed_size() -> usize;
+        fn roc__forHost_0_caller(flags: *const (), captures: *mut u8, model: *mut u8);
+        fn roc__forHost_0_size() -> usize;
+    }
+
+    unsafe {
+        let captures_size = roc__forHost_1_exposed_size();
+        let mut captures = RocList::from_raw_parts(
+            roc_alloc(captures_size, 0) as *mut u8,
```
Comment (bhansconnect) [inline crates/roc_host/src/roc.rs:908]: I get that alignment doesn't truly matter cause we don't use it currently, but an alignment of `0` is definitely wrong here.

File: crates/roc_host/src/roc.rs [hunk 3]
```diff
@@ -945,3 +886,89 @@ pub extern "C" fn roc_fx_tempDir() -> RocList<u8> {
...
+            captures_size,
+            captures_size,
+        );
+
+        let model_size = roc__forHost_0_size();
+        let mut model =
+            RocList::from_raw_parts(roc_alloc(model_size, 0) as *mut u8, model_size, model_size);
```
Comment (bhansconnect) [inline crates/roc_host/src/roc.rs:915]: I'm kinda amazed this works. `model` is not a list, it is a pointer. That said, I'm not sure it is correct, would need to think about this more. Given we box the model, the output size should always be the size of a single pointer.

**Comment by bhansconnect** [comment]
I think the core platform stuff and model loading should be correct now.

**Comment by bhansconnect** [comment]
Oh, though the bench script still isn't working. Intriguing.

**Comment by bhansconnect** [comment]
@Anton-4 how does one update the example in the readme when it depends on the url of a release. Since the platform got updated, we have to release a new version of basic-webserver for the example to point to.

**Comment by bhansconnect** [comment]
Otherwise, I think this is good to merge

File: crates/simple_bench/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    Command::new("roc")
+        .arg("build.roc")
+        .status()
+        .expect("Failed to build the host");
+
+    // Start the server
+    let mut server_process = {
+        Command::new("roc")
+            // .arg("--optimize")
+            .arg("--profiling")
```
Comment (Anton-4) [inline crates/simple_bench/src/main.rs:27]: Should this be `--optimize` so we get the best performance?

File: crates/simple_bench/src/main.rs
Comment (Anton-4) [inline crates/simple_bench/src/main.rs:1]: Should we leave this file out and recommend doing `wrk -t 4 -c 64 -d 20 http://127.0.0.1:8000` for any benchmarking?

Comment (bhansconnect) [reply crates/simple_bench/src/main.rs:1]: Probably. I think `simple_bench` ended up being more of a testing tool. I think there are lots of good web benchmarking tools that users can research and use. `wrk`, `wrk2`, `ab`, etc. Each with their own color of tradeoffs. Also, `wrk` functions great for testing as well. So I don't think `simple_bench` is worth adding.

Comment (bhansconnect) [reply crates/simple_bench/src/main.rs:1]: I'll write up something basic in the readme on benchmarking

Comment (Anton-4) [reply crates/simple_bench/src/main.rs:1]: Sounds good, thanks :)

File: examples/command.roc [hunk 5]
```diff
@@ -1,13 +1,19 @@
...
 import pf.Utc
 
-main : Request -> Task Response []
-main = \req ->
+Model : {}
 
+server = { init, respond }
+
+init : Task Model [Exit I32 Str]_
+init = Task.ok {}
```
Comment (Anton-4) [inline examples/command.roc:13]: It's probably best if we leave this for another PR but I'd like to minimize the boiler plate here some time. Perhaps we could make server a tag union like `[NoInit (Http.Request -> Task Http.Response [ServerErr Str]_), WithInit { init: _, respond: _}]`

**Comment by Anton-4** [comment]
I'll finish up my review on monday

File: examples/command.roc (hunk 5 above)
Comment (bhansconnect) [reply examples/command.roc:13]: I was looking at this change and it makes some of the state management and juggling quite a bit more complex in the platform. I would prefer to delay this to a later PR.

Also, this is minimal boilerplate currently.
```
Model : {}
server = { init: Task.ok {}, respond }

respond : Request, Model -> Task Response [ServerErr Str]_
respond = \req, _ ->
    # Log request datetime, method and url
    datetime = Utc.now! |> Utc.toIso8601Str

    Stdout.line! "$(datetime) $(Http.methodToStr req.method) $(req.url)"

    Task.ok { status: 200, headers: [], body: Str.toUtf8 "<b>Hello, world!</b></br>" }
```

Comment (lishaduck) [reply examples/command.roc:13]: Sorry to "interrupt"

What about making a `createServer` function that uses optional record params? It'd allow omitting the `init`, and potentially let you hide the implementation a bit (if you put it behind an opaque type, otherwise it's just like a JS-style `defineConfig`).
Slightly less verbose, about what it is now, but scales (just add an init).

Maybe I'm missing something, but it seems to me like that'd work.

Comment (bhansconnect) [reply examples/command.roc:13]: So part of the complication is that the app defines the model and the platform doesn't know what type it is.

So the init function is a `Task (Box Model) I32`. Having a default branch in the platform that generates a `Task.ok {}` for the model, will not unify if the model is not a `{}`. So the types are definitely a bit more finicky here.

I need to mess around with this more though. I'm sure there must be a clean way to make this work.

Comment (lishaduck) [reply examples/command.roc:13]: Ah, yeah, that makes sense. I figured the user's model might mess things up. I don't think the type system would let you do this, but in my head you'd be able to default the model to `{}` and force it define init if it's not `{}`, as you'd need to define it if it's not nothing anyway, right?
I don't _think_ you can do it, but that's the API I imagine people would like. It's what I'd like, at least.

Comment (bhansconnect) [reply examples/command.roc]: Also inconvenient, `Model` has to be specified no matter what. Roc doesn't have a way for us to optionally require a type be passed from the app to the platform. So even if we have a config function, it would be at a minimum something like:

```roc
Model : {}

server = noInit \req ->
    # Log request datetime, method and url
    datetime = Utc.now! |> Utc.toIso8601Str

    Stdout.line! "$(datetime) $(Http.methodToStr req.method) $(req.url)"

    Task.ok { status: 200, headers: [], body: Str.toUtf8 "<b>Hello, world!</b></br>" }
```

vs:
```roc
Model : {}

server = {
    init: Task.ok {},
    respond: \req, _ ->
        # Log request datetime, method and url
        datetime = Utc.now! |> Utc.toIso8601Str

        Stdout.line! "$(datetime) $(Http.methodToStr req.method) $(req.url)"

        Task.ok { status: 200, headers: [], body: Str.toUtf8 "<b>Hello, world!</b></br>" },
}
```

So a tiny bit denser, but not a ton. Though, I definitely wouldn't nest respond in the server record directly. I don't think it is great for readabilty. So more realistically, it would be:
```roc
Model : {}

server = { init: Task.ok {}, respond }
respond = \req, _ ->
    # Log request datetime, method and url
    datetime = Utc.now! |> Utc.toIso8601Str

    Stdout.line! "$(datetime) $(Http.methodToStr req.method) $(req.url)"

    Task.ok { status: 200, headers: [], body: Str.toUtf8 "<b>Hello, world!</b></br>" }
```

So only one extra line of code.

Comment (lishaduck) [reply examples/command.roc]: Yeah. At that point, it's not worth it, _unless_ you want to be able to add subscriptions/whateverelseElmhas in a non-breaking manner, but at that point, you're probably breaking other things anyway.

Comment (bhansconnect) [reply examples/command.roc]: > I don't think you can do it, but that's the API I imagine people would like. It's what I'd like, at least.

Yeah, there is probably a way to make the api work and be nicer, but I haven't quite found it yet. At least not without adding a bunch of extra conditionals to the platform.

Like we can make the `server` be a tag union `[ WithInit { init, respond }, NoInit respond ]`, but the users still has to specify a `Model` due to how roc works. Also, the platform has to do a lot of extra conditionals to make the types happy. It can no longer use `Model` directly. It has to use something like `[Model Model, NoModel]`. Just adds conditionals that feel very unnecessary.

Comment (lishaduck) [reply examples/command.roc]: Yeah, and the tag union looks ugly. I'd just leave it be, it can be improved some other time.

Though, why can't you extract model out? i.e., rather than passing it, just make the defineConfig-y function take a typeparam, then extract the typeparam. Or does the typesystem not allow that?

Comment (bhansconnect) [reply examples/command.roc]: I feel like I did something similar in the past, but I wasn't having luck when I attempted to do so today. Maybe I am just formulating something wrong.

File: README.md [hunk 6]
```diff
@@ -43,3 +46,25 @@ If you'd like to contribute, check out our [group chat](https://roc.zulipchat.co
 ## Developing / Building Locally
 
 If you have cloned this repository and want to run the examples without using a packaged release (...tar.br), you will need to build the platform first by running `roc build.roc`. Run examples with `roc examples/hello.roc` (on linux, add `--linker=legacy`).
+
+## Benchmarking
+
+Basic webserver should have decent performance due to being built on top of Rust's [hyper](https://hyper.rs).
+That said, it has a few known issues that hurt performance:
+1. We do [extra data copying on every request](https://github.com/roc-lang/basic-webserver/issues/23).
+2. Until roc has effect interpreters, basic-webserver can only do blocking io for effects. To work around this, every request is spawned in a blocking thread.
```
Comment (lukewilliamboswell) [inline README.md:55]: We should link to or reference "effect interpreters" here.

File: README.md [hunk 7]
```diff
@@ -43,3 +46,25 @@ If you'd like to contribute, check out our [group chat](https://roc.zulipchat.co
 ## Developing / Building Locally
 
 If you have cloned this repository and want to run the examples without using a packaged release (...tar.br), you will need to build the platform first by running `roc build.roc`. Run examples with `roc examples/hello.roc` (on linux, add `--linker=legacy`).
+
+## Benchmarking
```
Comment (lukewilliamboswell) [inline README.md:50]: Love the addition of this. Thank you for adding this @bhansconnect

File: platform/main.roc [hunk 8]
```diff
@@ -20,12 +23,74 @@ platform "webserver"
     packages {}
     imports [
         Task.{ Task },
+        Stderr.{ line },
     ]
-    provides [mainForHost]
+    provides [forHost]
```
Comment (lukewilliamboswell) [inline platform/main.roc:28]: I changed it to `forHost` here because I think it is useful to be distinct from the other platforms that have only one `main` entry point, and clarify that this is what roc is providing for the host. Hopefully as this will be an example platform people look at for ideas, we can move away from only having "mainForHost" in all of the platforms.

File: platform/main.roc [hunk 9]
```diff
@@ -20,12 +23,74 @@ platform "webserver"
...
+
+                    Tip: If you do not want to exit on this error, use `Task.mapErr` to handle the error.
+                    Docs for `Task.mapErr`: <https://www.roc-lang.org/packages/basic-cli/Task#mapErr>
+                    """
+                |> Task.onErr \_ -> Task.err 1
+                |> Task.await \_ -> Task.err 1
+
+respond : InternalHttp.RequestToAndFromHost, Box Model -> InternalTask.Task InternalHttp.ResponseToHost []
+respond = \request, boxedModel ->
+    when server.respond (InternalHttp.fromHostRequest request) (Box.unbox boxedModel) |> Task.result! is
```
Comment (lukewilliamboswell) [inline platform/main.roc:69]: For reference the error handling here was inspired from [this zulip discussion](https://roc.zulipchat.com/#narrow/stream/383402-API-Design/topic/Default.20error.20handling.20in.20basic-webserver/near/436375969).

Basically any unhandled Err's returned to the platform will be logged to stderr, and a http 500 error returned. 

This will be a nicer developer experience, particularly for quick demos or beginners who may not want to explicitly handle every error.

**Comment by lukewilliamboswell** [comment]
I concur with Brendan, this PR is good to merge. I'll wait for Anton to finish his review next week.

**Review by Anton-4 — APPROVED** [review]
Thanks @lukewilliamboswell and @bhansconnect :heart:
